packages/interactivity-router/src/assets/scripts.ts

```diff
@@ -0,0 +1,85 @@
+/**
+ * Internal dependencies
+ */
+import {
+	importPreloadedModule,
+	preloadWithMap,
+	type ModuleLoad,
+} from './dynamic-importmap';
+
+/**
+ * Script element containing the initial page's import map.
+ */
+const initialImportMapElement =
+	window.document.querySelector< HTMLScriptElement >(
+		'script#wp-importmap[type=importmap]'
+	);
+
+/**
+ * Data from the initial page's import map.
+ *
+ * Pages containing any of the imports present on the original page
+ * in their import maps should ignore them, as those imports would
+ * be handled natively.
+ */
+const initialImportMap = initialImportMapElement
+	? JSON.parse( initialImportMapElement.text )
+	: { imports: {}, scopes: {} };
+
+/**
+ * IDs of modules that should be resolved by the browser rather than
+ * processed internally.
+ */
+const modulesResolvedNatively = new Set< string >();
+
+/**
+ * Mark the specified module as natively resolved.
+ * @param id Module ID.
+ */
+export const setModuleAsNativelyResolved = ( id: string ) => {
+	modulesResolvedNatively.add( id );
+};
+
+/**
+ * Resolve and fetch modules present in the passed document, using the
+ * document's import map to resolve them.
+ *
+ * @param doc Document containing the modules to preload.
+ * @return Array of promises that resolve to a `ModuleLoad` instance.
+ */
+export const preloadModules = ( doc: Document ) => {
+	// Extract the import map from the document.
+	const importMapElement = doc.querySelector< HTMLScriptElement >(
+		'script#wp-importmap[type=importmap]'
+	);
+	const importMap = importMapElement
+		? JSON.parse( importMapElement.text )
+		: { imports: {}, scopes: {} };
+
+	// Remove imports also in the initial page's import map.
+	// Those should be handled natively.
+	for ( const key in initialImportMap.imports ) {
+		delete importMap.imports[ key ];
+	}
+
+	// Get the URL of all modules contained in the document.
+	const moduleUrls = [
+		...doc.querySelectorAll< HTMLScriptElement >(
+			'script[type=module][src]'
```

Comment:
I think we should do an opt-in here, at least initially.

```html
<script type="module" src="..." data-interactivity-api-compat></script>
```

And we mark only the view script modules that have `supports.interactivity` equal to true.

But we don't need to do it as part of this PR.